### PR TITLE
receiver/azureblobreceiver First PR

### LINKED ISSUE #8834: New component: Azure Blob Receiver
## The purpose and use-cases of the new component
This receiver reads logs and traces from [Azure Blob Storage](https://azure.microsoft.com/services/storage/blobs/).

Correlates with #8823 

## Example configuration for the component
The following settings are required:

- `connection_string` (no default): Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
- `event_hub_end_point` (no default): Azure Event Hub endpoint triggering on the `Blob Create` event 

The following settings can be optionally configured:

- `logs_container_name` (default = "logs"): Name of the blob container with the logs
- `traces_container_name` (default = "traces"): Name of the blob container with the traces

Example:

```yaml
receivers:
  azureblob:
    connection_string: DefaultEndpointsProtocol=https;AccountName=accountName;AccountKey=+idLkHYcL0MUWIKYHm2j4Q==;EndpointSuffix=core.windows.net
    event_hub_end_point: Endpoint=sb://oteldata.servicebus.windows.net/;SharedAccessKeyName=otelhubbpollicy;SharedAccessKey=mPJVubIK5dJ6mLfZo1ucsdkLysLSQ6N7kddvsIcmoEs=;EntityPath=otellhub    
```

The receiver subscribes [on the events](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-event-overview) published by Azure Blob Storage and handled by Azure Event Hub. When it receives `Blob Create` event, it reads the logs or traces from a corresponding blob and deletes it after processing.

## Telemetry data types supported
Logs, Traces

## Sponsor
Sponsor is needed

### PR DESCRIPTION
**Pull request by eedorenko**
Description:
azureblobreceiver reads logs and traces from Azure Blob Storage.

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8834

Documentation:
Readme describing receiver functionality and configuration.

### CODE REVIEW

**Review by codeboten — COMMENTED** [review]
@pmm-sumo please review

**Comment by codeboten** [comment]
pmm-sumo is no longer involved in the project. @mx-psi as the next sponsor on the list, please review this component.

File: receiver/azureblobreceiver/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Azure Blob Receiver
+
+This receiver reads logs and trace data from [Azure Blob Storage](https://azure.microsoft.com/services/storage/blobs/).
+
+## Configuration
+
+The following settings are required:
+
+- `connection_string` (no default): Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
+- `event_hub_end_point` (no default): Azure Event Hub endpoint triggering on the `Blob Create` event
```
Comment (mx-psi) [inline receiver/azureblobreceiver/README.md:10]: We usually use just `endpoint` for this. If there could be several endpoints then I would use:

```
event_hub:
  endpoint:
```

File: receiver/azureblobreceiver/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+The following settings are required:
+
+- `connection_string` (no default): Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
+- `event_hub_end_point` (no default): Azure Event Hub endpoint triggering on the `Blob Create` event 
+
+The following settings can be optionally configured:
+
+- `logs_container_name` (default = "logs"): Name of the blob container with the logs
+- `traces_container_name` (default = "traces"): Name of the blob container with the traces
```
Comment (mx-psi) [inline receiver/azureblobreceiver/README.md:15]: Do you think more options specific for logs or traces could be added in the future? If so, I would create sections for these:

```
logs:
  container_name: logs
traces:
  container_name: traces
```

File: receiver/azureblobreceiver/factory.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	receiver, err := f.getReceiver(set, cfg)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return receiver, nil
+}
+
+func (f *blobReceiverFactory) getReceiver(
```
Comment (mx-psi) [inline receiver/azureblobreceiver/factory.go:93]: I believe the way we usually do this is with this module: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/sharedcomponent/sharedcomponent.go You can take a look at how the OTLP receiver does this here: https://github.com/open-telemetry/opentelemetry-collector/blob/27fdc2e2ded2b0900d165edebbf4c26ffc83c2d0/receiver/otlpreceiver/factory.go#L90-L92

File: receiver/azureblobreceiver/go.mod [hunk 4]
```diff
@@ -0,0 +1,24 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/azureblobreceiver
+
+go 1.17
+
+require go.opentelemetry.io/collector v0.45.1-0.20220222205432-ec4f7040191d
```
Comment (mx-psi) [inline receiver/azureblobreceiver/go.mod:5]: We need to move this to the latest released version, `v0.56.0`

**Review by mx-psi — CHANGES_REQUESTED** [review]

File: go.mod [hunk 5]
```diff
@@ -825,8 +825,6 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfi
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver => ./receiver/awsxrayreceiver
 
-replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/azureeventhubreceiver => ./receiver/azureeventhubreceiver
```
Comment (mx-psi) [inline go.mod:828]: I don't understand the relation of this change with the rest of the PR

File: receiver/azureblobreceiver/factory.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	r := f.receivers.GetOrAdd(cfg, func() component.Component {
+		receiverConfig, ok := cfg.(*Config)
+
+		if !ok {
+			err = errUnexpectedConfigurationType
+			return nil
+		}
+
+		var receiver component.Receiver
+		receiver, err = NewReceiver(*receiverConfig, set)
```
Comment (mx-psi) [inline receiver/azureblobreceiver/factory.go:112]: You should handle this error

File: receiver/azureblobreceiver/receiver.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+func (b *blobReceiver) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+func (b *blobReceiver) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+// Returns a new instance of the log receiver
+func NewReceiver(config Config, set component.ReceiverCreateSettings) (component.Receiver, error) {
```
Comment (mx-psi) [inline receiver/azureblobreceiver/receiver.go:35]: Does this need to be public?

File: receiver/azureblobreceiver/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+This receiver reads logs and trace data from [Azure Blob Storage](https://azure.microsoft.com/services/storage/blobs/).
+
+## Configuration
+
+The following settings are required:
+
+- `connection_string:` (no default): Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
+- `event_hub:`
+  `  end_point:` (no default): Azure Event Hub endpoint triggering on the `Blob Create` event
```
Comment (mx-psi) [inline receiver/azureblobreceiver/README.md:18]: For consistency with existing components this should be `endpoint`, not `end_point`

**Review by mx-psi — COMMENTED** [review]
Looking good, just a few more comments

File: receiver/azureblobreceiver/factory.go (hunk 6 above)
Comment (eedorenko) [reply receiver/azureblobreceiver/factory.go:112]: It is handled [here](https://github.com/eedorenko/opentelemetry-collector-contrib/blob/0744f67d2c2e9c0470a72faad98580f31a283197/receiver/azureblobreceiver/factory.go#L116)

File: go.mod (hunk 5 above)
Comment (eedorenko) [reply go.mod]: There is no relation. Accidentally.

File: receiver/azureblobreceiver/factory.go (hunk 6 above)
Comment (mx-psi) [reply receiver/azureblobreceiver/factory.go]: Oh, okay, thanks. I got confused :)

File: receiver/azureblobreceiver/config.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azureblobreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/azureblobreceiver"
+
+import (
+	"go.opentelemetry.io/collector/config"
+)
+
+type Config struct {
```
Comment (mx-psi) [inline receiver/azureblobreceiver/config.go:21]: Can you open a follow up PR to document the fields directly in the Go struct? This is not urgent, but will be useful in the future for automatic configuration schema generation

**Review by mx-psi — APPROVED** [review]
LGTM

**Comment by mx-psi** [comment]
You need to add this component to `.github/CODEOWNERS`. You can list yourself and also me as the sponsor.

```
FAIL: "receiver/azureblobreceiver" not included in CODEOWNERS
```

File: .github/CODEOWNERS [hunk 10]
```diff
@@ -136,6 +136,7 @@ receiver/awscontainerinsightreceiver/                @open-telemetry/collector-c
 receiver/awsecscontainermetricsreceiver/             @open-telemetry/collector-contrib-approvers @Aneurysm9
 receiver/awsfirehosereceiver/                        @open-telemetry/collector-contrib-approvers @Aneurysm9
 receiver/awsxrayreceiver/                            @open-telemetry/collector-contrib-approvers @willarmiros
+receiver/azureblobreceiver/                          @open-telemetry/collector-contrib-approvers @eedorenko @pmm-sumo
```
Comment (mx-psi) [inline .github/CODEOWNERS:139]: Przemek is no longer involved in the project:

```suggestion
receiver/azureblobreceiver/                          @open-telemetry/collector-contrib-approvers @eedorenko @mx-psi
```

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
Looks good, will merge once all tests pass (probably tomorrow morning first thing if nobody else has done it :))

**Pull request merged**